New output buffer implementation

dain opened the pull request.

dain commented:
Note: Github, in its infinite wisdom, has shuffled the commit order.

dain commented:
@haozhun can you review "Add new PartitionedBuffer" and "Add new BroadcastBuffer"

haozhun commented:
Don't act on my reviews until I say so. I tend to modify my comments a lot. I wish github comments can be staged.

* Move SharedBuffer and related classes to new package
 * I assume nothing changed except deletion of a file and move files. Looks good.
* Add new PartitionedBuffer
* Move shared buffer not-full notification to a new thread
* Add buffer type to OutputBuffers to allow for different buffer implementations
* Add LazyBuffer to delay creation until plan is sent
* Register output buffers before adding splits
* Declare out buffers immediately for non-broadcast stages
* Replace TaskId in output buffers with a generic OutputBufferId
* Change TaskId.id to be an int
* Fix thread safty issue in BroadcastOutputBufferManager
* Extract interface OutputBuffer from SharedBuffer
* Add OutputBuffers.createInitialEmptyOutputBuffers
* Move SharedBuffer and related classes to new package

haozhun commented on presto-main/src/main/java/com/facebook/presto/OutputBuffers.java:
```diff
@@ -31,7 +31,11 @@
+
+    public static OutputBuffers createInitialEmptyOutputBuffers()
+    {
+        return new OutputBuffers(0, false, ImmutableMap.<TaskId, Integer>of());
```
`ImmutableMap.of()` will just work

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBuffer.java:
```diff
@@ -0,0 +1,91 @@
+    void setOutputBuffers(OutputBuffers newOutputBuffers);
+
+    /**
+     * Gets pages from the output buffer, and acknowledges all pages received from the request
```
I can't read/understand the `token` part of this javadoc. Rephrase?

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBuffer.java:
```diff
@@ -0,0 +1,91 @@
+    ListenableFuture<?> enqueue(int partition, Page page);
+
+    /**
+     * Disallows adding more pages to this output.
```
I believe the intent of this call is to declare, not to enforce. The enforcement is just a sanity check?

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBuffer.java:
```diff
@@ -0,0 +1,91 @@
+    void destroy();
+
+    /**
+     * Fail the buffer, discarding all pages, but blocking readers.
```
It would be helpful to add additional comments to explain what happens after readers are blocked.

haozhun commented on presto-main/src/main/java/com/facebook/presto/OutputBuffers.java:
```diff
@@ -34,19 +35,19 @@
     public static OutputBuffers createInitialEmptyOutputBuffers()
     {
-        return new OutputBuffers(0, false, ImmutableMap.<TaskId, Integer>of());
+        return new OutputBuffers(0, false, ImmutableMap.<OutputBufferId, Integer>of());
```
`<...>` not needed

haozhun commented on presto-main/src/test/java/com/facebook/presto/execution/scheduler/TestSourcePartitionedScheduler.java:
```diff
@@ -458,7 +458,7 @@ private SqlStageExecution createSqlStageExecution(StageExecutionPlan tableScanPl
                 locationFactory.createStageLocation(stageId),
                 tableScanPlan.getFragment(),
-                new MockRemoteTaskFactory(executor),
+                createInitialEmptyOutputBuffers(), new MockRemoteTaskFactory(executor),
```
formatting

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java:
```diff
@@ -217,8 +224,8 @@ public synchronized void addExchangeLocations(PlanFragmentId fragmentId, Set<URI
-            for (URI exchangeLocation : exchangeLocations) {
-                newSplits.put(remoteSource.getId(), createRemoteSplitFor(task.getTaskInfo().getTaskId(), exchangeLocation));
+            for (BiFunction<TaskId, Integer, URI> exchangeLocation : exchangeLocations) {
+                newSplits.put(remoteSource.getId(), createRemoteSplitFor(task.getTaskId(), task.getPartition(), exchangeLocation));
```
Do we need to have both `taskId` and `partition` here? When are they different?

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -192,10 +197,27 @@ else if (queryStateMachine.getQueryState() == QueryState.STARTING) {
+        // Non-broadcast distributions declare output buffers immediately
+        OutputBuffers initialOutputBuffers = createInitialEmptyOutputBuffers();
+        if (!plan.getFragment().getPartitionFunction().getPartitioningHandle().equals(FIXED_BROADCAST_DISTRIBUTION)) {
+            int partitionCount = IntStream.of(plan.getFragment().getPartitionFunction().getBucketToPartition().get()).max().getAsInt() + 1;
```
`Ints.max` from guava

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -467,18 +488,28 @@ public void processScheduleResults(StageState newState, Set<RemoteTask> newTasks
+                for (RemoteTask task : newTasks) {
+                    // if this stage has a broadcast output buffer...
+                    URI taskLocation = task.getTaskInfo().getSelf();
+                    if (currentStageIsBroadcast) {
```
A possible alternative to this:

* Keep both task id and partition id in the URL
* Let the OutputBuffer pick the one (of task id and partition id) it is interested in and throw out the other one

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -467,18 +488,28 @@ public void processScheduleResults(StageState newState, Set<RemoteTask> newTasks
-                parent.get().addExchangeLocations(currentStageFragmentId, newExchangeLocations, noMoreTasks);
+                ImmutableList.Builder<BiFunction<TaskId, Integer, URI>> newExchangeLocations = ImmutableList.builder();
+                for (RemoteTask task : newTasks) {
+                    // if this stage has a broadcast output buffer...
```
This comment is in the wrong place, and doesn't seem informative

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -192,10 +197,27 @@ else if (queryStateMachine.getQueryState() == QueryState.STARTING) {
+
+        // Non-broadcast distributions declare output buffers immediately
+        OutputBuffers initialOutputBuffers = createInitialEmptyOutputBuffers();
+        if (!plan.getFragment().getPartitionFunction().getPartitioningHandle().equals(FIXED_BROADCAST_DISTRIBUTION)) {
```
Alternatively, you could do the initial output buffer declaration inside constructor of PartitionOutputBufferManager, and do nothing in `addOutputBuffers` (except maybe sanity check)

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/LazyBuffer.java:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    @Override
+    public void abort(OutputBufferId bufferId)
```
Question:

* How do I search for usage a REST endpoint inside Presto?
* I thought `abort` is used to ack that the last page is received. I suppose I'm wrong. Otherwise, `abort` can't be called before `delegate` is set.

haozhun replied:
This function should check `state`

dain replied:
The state doesn't matter here.  If the delegate doesn't exist and the lazy buffer is done, we only save a single string value.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/SharedBuffer.java:
```diff
@@ -104,6 +104,13 @@ public SharedBuffer(TaskId taskId, String taskInstanceId, Executor executor, Dat
         this.memoryManager = new SharedBufferMemoryManager(maxBufferSize.toBytes(), systemMemoryUsageListener);
     }
 
+    public SharedBuffer(String taskInstanceId, StateMachine<BufferState> state, DataSize maxBufferSize, SystemMemoryUsageListener systemMemoryUsageListener)
```
I would delegate the original constructor to this new one.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/LazyBuffer.java:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    @Override
+    public CompletableFuture<BufferResult> get(OutputBufferId bufferId, long startingSequenceId, DataSize maxSize)
```
It seems this variable is sometimes called `startingSequencedId`, and sometimes `token`.

haozhun replied:
This function should check `state`

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/LazyBuffer.java:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
```
What happens if this occurs after `fail` or `destroy` has been called.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/LazyBuffer.java:
```diff
@@ -0,0 +1,297 @@
+    @Override
+    public void destroy()
+    {
+        OutputBuffer outputBuffer;
```
I suppose this should be ignored if it is already in a terminal state.

haozhun commented on presto-main/src/main/java/com/facebook/presto/OutputBuffers.java:
```diff
@@ -75,6 +106,13 @@ public boolean isNoMoreBufferIds()
+        checkArgument(type == newOutputBuffers.getType(), "newOutputBuffers has a different type");
+
+        if (noMoreBufferIds) {
+            checkArgument(this.equals(newOutputBuffers), "Expected buffer to not change after no more buffers is set");
```
I think the original order make sense. By moving this check early, it can breaks when `transitions` were received out of order.

haozhun replied:
I just noticed that all callers check the version between calling `checkValidTransition`. This function would just be ignored if the new version is no newer than the original one. Making the following version related check meaningless.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/LazyBuffer.java:
```diff
@@ -65,6 +66,7 @@ public LazyBuffer(TaskId taskId, String taskInstanceId, Executor executor, DataS
         requireNonNull(taskId, "taskId is null");
         requireNonNull(executor, "executor is null");
         this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceId is null");
+        this.executor = requireNonNull(executor, "executor is null");
```
`requireNonNull` check is duplicate

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+                    this.pendingRead = null;
+
+                    // if buffer is finished return an empty page
+                    // this could be a request for a buffer that never existed, but that is ok since the buffer
```
I don't understand

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+                return resultFuture;
+            }
+
+            public void abort()
```
I would inline this function. `abort` as a name is not clear to me when I read its usages. Additionally, one of the caller is wrong. It should say `true` instead of `false` for the last field. Removing `abort` will also enable removing `taskInstanceId` from this class, which doesn't really belong here.

haozhun replied:
Does it make more sense to return the current sequence id as the next sequence id?

dain replied:
This point of this method is to tell the reader to comeback to get more data.  I believe the case you are referring to is in `setNoMorePages`, and this is safe as the client will simply come back and then get the "finished" result. This is inefficient, so I will change that case.

dain replied:
The abort call, simply tells the client there is currently no data at that position, and it should try again (at the same position).  If there are concurrent requests, it is assumed there is a single client for the buffer and the client will recognize the response is for and abandoned request.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+                // if there are no pages, and there will be no more pages, then we are finished
+                // never finish a failed buffer
+                if (pages.isEmpty() && !state.get().canAddPages() && state.get() != FAILED) {
+                    finished.set(true);
```
Is this necessary given `destroy` will be called by `abort`?

dain replied:
The finished flag is generally an optimization, but is required to make the `getBufferInfo` method lock free.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+                    // Return results immediately if we have data or are already finished
+                    // Additionally, if the read is for data before the currently acknowledged pages, this is
+                    // likely an out of order request, so return an empty result.
+                    if (!pages.isEmpty() || finished.get() || sequenceId < currentSequenceId.get()) {
```
`sequenceId < currentSequenceId.get()` is not possible

haozhun replied:
`finished.get()` is not possible

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+                verify(currentSequenceId.compareAndSet(oldCurrentSequenceId, oldCurrentSequenceId + pagesToRemove));
+
+                bufferedBytes.addAndGet(-bytesRemoved);
+                verify(bufferedBytes.get() >= 0);
```
`verify(bufferedBytes.addAndGet(-bytesRemoved) >= 0)`

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+        {
+            this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceId is null");
+            this.partitionNumber = partitionNumber;
+            this.state = requireNonNull(state, "state is null");
```
It would be great if it's possible to remove `state` from `Partition`.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+            synchronized (this) {
+                pendingRead = this.pendingRead;
+                this.pendingRead = null;
+            }
```
Something needs to be set to quickly return future read requests?

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+            }
+        }
+
+        private void setNoMorePages()
```
`public`?

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedBuffer.java:
```diff
@@ -0,0 +1,543 @@
+        private final AtomicLong bufferedBytes = new AtomicLong();
+
+        private final AtomicLong currentSequenceId = new AtomicLong();
+        private final AtomicBoolean finished = new AtomicBoolean();
```
The meaning of `finished` is not apparent to me. It appears like all functions except `getInfo` should be no-ops once this flag is set. But then, the behavior of some functions here doesn't match my theory.

dain replied:
Can you be more specific about which functions you don't understand?

haozhun commented:
I modified my comments a lot. Please don't look at your email notifications.

Feel free to rebase. For commits that looks good, you can amend those commits. I'll not look at them again unless you feel it's needed and/or some of my comments for that commit doesn't make sense.



* Move SharedBuffer and related classes to new package
  * I assume nothing changed except deletion of a file and move files. Looks good.
* Add OutputBuffers.createInitialEmptyOutputBuffers
  * Looks good, except minor comments
* Extract interface OutputBuffer from SharedBuffer
  * Looks good, except comments for javadoc
* Fix thread safty issue in BroadcastOutputBufferManager
  * Typo in commit message
  * Looks good
* Change TaskId.id to be an int
  * Looks good
* Replace TaskId in output buffers with a generic OutputBufferId
  * Looks good
* Declare out buffers immediately for non-broadcast stages
  * This commit also changes URI for output buffers. The commit message should be revised.
  * I do have an alternative suggestion. I'm not sure whether it's a good idea to intialize `initialOutputBuffers` differently directly inside SqlQueryScheduler. To me, it feels like this naturally fits into the role of the `OutputBufferManager`.
  * I also brought up an alternative suggestion about URL, as we've previously discussed.
* Register output buffers before adding splits
  * why?
* Add LazyBuffer to delay creation until plan is sent
  * There may be some problems. See comments.
* Add buffer type to OutputBuffers to allow for different buffer implementations
  * Looks good
* Move shared buffer not-full notification to a new thread
  * Looks good, but I would like to have a quick chat to understand it more
* Add new PartitionedBuffer
  * See comments
* Add new BroadcastBuffer
  * Haven't looked because it has a lot in common with `PartitionedBuffer`. I'll review this after the comments on PartitionedBuffer are applied.

dain commented:
@haozhun I applied your suggestions to the existing commits, except for the changes to the new PartitionedBuffer.  Those changes are in a separate commit.  Also, I have not touched the BroadcastBuffer.  After you are happy with PartitionedBuffer, I will update it.

haozhun commented:
Looks like you missed these two comments:

* Fix thread safty issue in BroadcastOutputBufferManager
  * Typo in commit message
  * Looks good
* Declare out buffers immediately for non-broadcast stages
  * This commit also changes URI for output buffers. The commit message should be revised.
  * I do have an alternative suggestion. I'm not sure whether it's a good idea to intialize `initialOutputBuffers` differently directly inside SqlQueryScheduler. To me, it feels like this naturally fits into the role of the `OutputBufferManager`.
  * I also brought up an alternative suggestion about URL, as we've previously discussed.

dain commented:
I renamed the classes to make it clear which ones are specific to the old SharedBuffer code and which ones are generic.  I also named OutputBuffer implementations to end with "OutputBuffer" for clarity.
* Rename SharedBufferInfo to OutputBufferInfo
* Rename SharedBuffer to SharedOutputBuffer
* Rename SharedBufferMemoryManager to OutputBufferMemoryManager

I reworked the URI and BufferId management logic in the query scheduler.  The key insight was that I could force the partition number and the taskId to be the same for partitioned outputs, with some minor changes.  With that, I didn't need most of my changes, so the old code is restored:
* Guarantee task id and output partition are the same
* Remove unused partition from OutputBufferManager
* Remove unused partition from RemoteTask
* Declare out buffers immediately for non-broadcast stages

Finally, when working on the new BroadcastOutputBuffer, I discovered that I could reuse the internal PartitionedOutputBuffer.Partition object in BroadcastOutputBuffer by adding reference counting.  I pulled out this class into a ClientBuffer and added tests for the client protocol. I'll squash the first two commits:
* Extract PartitionedOutputBuffer.Partition into ClientBuffer 
* Add reference counting to ClientBuffer
* Add new BroadcastOutputBuffer

haozhun commented:
* Looks good
  * Move SharedBuffer and related classes to new package
  * Add OutputBuffers.createInitialEmptyOutputBuffers
  * Rename SharedBufferInfo to OutputBufferInfo
  * Rename SharedBuffer to SharedOutputBuffer
  * Rename SharedBufferMemoryManager to OutputBufferMemoryManager
  * Change TaskId.id to be an int
  * Replace TaskId in output buffers with a generic OutputBufferId
  * Add LazyOutputBuffer to delay creation until plan is sent
  * Add buffer type to OutputBuffers to allow for different buffer implementations
  * Extract PartitionedOutputBuffer.Partition into ClientBuffer

(Review In Progress)
* Extract interface OutputBuffer from SharedBuffer
  * Looks good, except comments for javadoc
  * `OutputBuffer.setNoMorePages` javadoc
  * `OutputBuffer.get`
```
     * Gets pages from the output buffer, and acknowledges all pages received from the last
     * request.  The initial token is zero. Subsequent tokens are acquired from the
     * next token field in the BufferResult returned from the previous request.
```
* Fix thread safty issue in BroadcastOutputBufferManager
  * Typo in commit message
  * Looks good
* Change TaskId.id to be an int
* Replace TaskId in output buffers with a generic OutputBufferId
* Guarantee task id and output partition are the same
* Remove unused partition from OutputBufferManager
* Remove unused partition from RemoteTask
* Declare out buffers immediately for non-broadcast stages
  * This commit also changes URI for output buffers. The commit message should be revised.
  * I do have an alternative suggestion. I'm not sure whether it's a good idea to intialize `initialOutputBuffers` differently directly inside SqlQueryScheduler. To me, it feels like this naturally fits into the role of the `OutputBufferManager`.
* Register output buffers before adding splits
  * Add more comments
* Move shared buffer not-full notification to a new thread
  * `LazyBuffer.LazyBuffer`: duplicate `requireNonNull` for `executor`
  * set `notFull` through the executor in `SharedBufferMemoryManager.setNoBlockOnFull`
* Add new PartitionedOutputBuffer
  * For `PartitionedBuffer.fail`, the "// DO NOT" comment should be inside the if body.
  * For `PartitionedBuffer.Partition.PendingRead.abort`, rename to something like `completeResultFutureWithEmpty`.
  * In `PartitionedBuffer.Partition.processRead(long, DataSize)`, verification of sequenceId == currentSequenceId should be before the code that returns a completed empty result for finished buffer.
  * `PartitionedBuffer.Partition.setNoMorePages` shouldn't be synchronized at method level.
* Add reference counting to ClientBuffer
* Add new BroadcastBuffer
  * Haven't looked because it has a lot in common with `PartitionedBuffer`. I'll review this after the comments on PartitionedBuffer are applied.

haozhun commented:
Separate discussion, (not entirely related to this pull request, be worth noting and related).

I believe current communication protocol for output buffer is inherently not thread safe:
* "consumer" send request A (sequence id = 3)
* "producer" receives request A 
* "producer" send response B (2 pages, next sequence id = 5)
* "consumer" retries request C (sequence id = 3)
* "producer" receives request C
* "consumer" receives response B
* "producer" send response D (4 pages, next sequence id = 7)
* "consumer" receives response D

Now, "consumer" got duplicate pages.

I believe this needs to be fixed after this PR is merged. There are basically 2 possible solutions:

* Let the client understand that the "token" is "sequence id"
* Include the request sequence id in responses.

dain commented:
```markdown
* Declare out buffers immediately for non-broadcast stages
  * PartitionedOutputBufferManager.addOutputBuffers
    * Validate that noMoreBuffer does not change from `true` to `false`, unless there's a valid use case to do that.
    * Calling `withBuffer` a bunch of times will allocate a bunch of objects and create the `version` a lot. I guess these don't really matter. But I would create a map and then pass the map in just to be on the safe side.
```
The `noMoreBuffer` check is handled by the validate code, and the object allocation is not really a problem when you consider that `Map` will allocate an object for each entry anyway (also partitions is typically small).

The real issues is with the version number which the `checkValidTransition` code validates.  I just rewrote the code to check the buffers directly and avoided the `checkValidTransition` code.

```markdown
* Move shared buffer not-full notification to a new thread
  * Question about `OutputBufferMemoryManager`:
    * Is there any particular reason `updateMemoryUsage` submit the job outside the sync block and `setNoBlockOnFull` submit the job inside the sync block?
```
No, I simplified the code.

```markdown
* Add new PartitionedOutputBuffer
  * Why is calling `checkFlushComplete` necessary/useful in `setNoMorePages` or the constructor?
```
`setNoMorePages` can free a reader which will see the finished flag.  Then there is a race to destroy the buffers.  It is simpler to just double check at the end of the method, then try to determine if this is a benign race.  As for the constructor, it is likely not needed, but doesn't hurt anything (potentially, you could have zero partitions, but that seems weird).

```markdown
* Add reference counting to ClientBuffer
  * In `destroy`, `pendingRead` should be completed with `emptyResults(taskInstanceId, sequenceId, true)`
```
This doesn't matter.  There is only one client and the destroy comes from the client.  This has the benefit of keeping the API surface area smaller.

dain commented:
I will merge once the current release goes out
